File: cpp/hslm.cpp
#include "excavator2/hslm.hpp"
#include <cmath>
#include <vector>

namespace excavator2::hslm {
namespace {
// Literal ELNSUM from FastJointSLMLibraryI.f; retain its evaluation order.
double elnsum(double x, double y) {
    return x > y ? x + std::log(1 + std::exp(y-x)) : y + std::log(1 + std::exp(x-y));
}
}
void segment(std::size_t n, std::size_t k, const double* values, const double* means,
             double mi, double smu, double sepsilon, const double* eta,
             const double* initial, std::int32_t* path, double* transitions,
             double* emissions, double* scores, std::int32_t* predecessors) {
    std::vector<double> g(k), previous(k), current(k), block(k*k);
    std::vector<std::int32_t> history(n*k, 0);
    for (std::size_t j=0; j<k; ++j) {
        const double difference = means[j]-mi;
        g[j] = -(difference*difference/(2*(smu*smu)));
    }
    double norm=g[0];
    for (std::size_t j=1; j<k; ++j) norm=elnsum(norm,g[j]);
    for (auto& value:g) value-=norm;
    // The unsuffixed Fortran PI literal is rounded to REAL*4, then promoted.
    const double constant=std::log(1/(std::sqrt(2*3.1415927410125732421875)*sepsilon));
    for (std::size_t t=0; t<n; ++t) {
        if (t>0) {
            const double jump=std::log(eta[t-1]), stay=std::log(1-eta[t-1]);
            for (std::size_t source=0; source<k; ++source) {
                for (std::size_t dest=0; dest<k; ++dest) {
                    const double change=jump+g[dest];
                    block[source*k+dest]=source==dest ? elnsum(stay,change) : change;
                    if (transitions) transitions[(t-1)*k*k+source*k+dest]=block[source*k+dest];
                }
            }
        }
        for (std::size_t j=0; j<k; ++j) {
            const double z=(values[t]-means[j])/sepsilon;
            const double emission=constant+(-0.5*(z*z));
            if (emissions) emissions[t*k+j]=emission;
            if (t==0) current[j]=initial[j]+emission;
            else {
                double best=previous[0]+block[j];
                std::int32_t index=1;
                for (std::size_t source=1; source<k; ++source) {
                    const double candidate=previous[source]+block[source*k+j];
                    if (candidate>best) { best=candidate; index=static_cast<std::int32_t>(source+1); }
                }
                history[t*k+j]=index;
                current[j]=best+emission;
            }
            if (scores) scores[t*k+j]=current[j];
            if (predecessors) predecessors[t*k+j]=history[t*k+j];
        }
        previous.swap(current);
    }
    std::size_t best=0;
    for (std::size_t j=1; j<k; ++j) if (previous[j]>previous[best]) best=j;
    path[n-1]=static_cast<std::int32_t>(best+1);
    for (std::size_t t=n-1; t>0; --t) path[t-1]=history[t*k+path[t]-1];
}
}

File: src/excavator2/reference/hslm.py
"""Single-profile, literal O(T K²) reference for FastJointSLMLibraryI.f.

Matrices use Python orientation: transitions[t, source, destination],
emissions/scores/predecessors[t, state]. Paths and predecessors remain one-based.
The Fortran PI literal rounds to binary32 before promotion to binary64.
Keep that value and the original log-sum evaluation order.
"""

import math

import numpy as np


def elnsum(x: float, y: float) -> float:
    return x + math.log(1 + math.exp(y - x)) if x > y else y + math.log(1 + math.exp(x - y))


def matrices(values, means, mi, smu, sepsilon, eta):
    k = len(means)
    g = -((means - mi) ** 2 / (2 * smu**2))
    norm = float(g[0])
    for value in g[1:]:
        norm = elnsum(norm, float(value))
    g = g - norm
    transitions = np.empty((len(eta), k, k))
    for t, change in enumerate(eta):
        log_change = math.log(change)
        log_stay = math.log(1 - change) if change < 1 else -math.inf
        for source in range(k):
            for destination in range(k):
                jump = log_change + g[destination]
                transitions[t, source, destination] = (
                    elnsum(log_stay, jump) if source == destination else jump
                )
    emissions = np.empty((len(values), k))
    constant = math.log(1 / (math.sqrt(2 * 3.1415927410125732421875) * sepsilon))
    for t, value in enumerate(values):
        for j, mean in enumerate(means):
            emissions[t, j] = constant + -0.5 * ((value - mean) / sepsilon) ** 2
    return transitions, emissions


def viterbi(initial, transitions, emissions):
    n, k = emissions.shape
    scores = np.empty((n, k))
    predecessors = np.zeros((n, k), dtype=np.int32)
    scores[0] = initial + emissions[0]
    for t in range(1, n):
        for j in range(k):
            best = scores[t - 1, 0] + transitions[t - 1, 0, j]
            index = 0
            for source in range(1, k):
                candidate = scores[t - 1, source] + transitions[t - 1, source, j]
                if candidate > best:
                    best, index = candidate, source
            predecessors[t, j] = index + 1
            scores[t, j] = best + emissions[t, j]
    path = np.empty(n, dtype=np.int32)
    path[-1] = np.argmax(scores[-1]) + 1
    for t in range(n - 2, -1, -1):
        path[t] = predecessors[t + 1, path[t + 1] - 1]
    return path, scores, predecessors
